File: src/countAndSay/Sol.kt
package countAndSay

object Solution {
    fun countAndSay(n: Int): String {
        tailrec fun rle(text: String, prev: String = ""): String {
            if (text.isEmpty()) return prev
            val cur = text[0]
            val cnt = text.takeWhile{ it == cur }.count()
            return rle(text.substring(cnt),prev + "$cnt$cur" )
        }
        tailrec fun sayCntDigit(n: Int, acc: String = "1"): String =
            if (n == 1) acc
            else sayCntDigit(n-1, rle(acc))
        return sayCntDigit(n)
    }
}

File: src/longestCommonPrefix/Sol.scala
package longestCommonPrefix

object Sol {
  def longestCommonPrefix(strs: Array[String]): String =
    if (strs.isEmpty) ""
    else if (strs.length == 1) strs(0)
    else strs.reduce(calcPrefix)

  private def calcPrefix(s1: String, s2: String): String =
    s1.substring(0,s1.lazyZip(s2).takeWhile(p => p._1 == p._2).count(_ => true))
}
